File: src/com/hard/BinaryTreePostorderTraversal_145.java
/**
 * 
 */
package com.hard;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Stack;

/**
 * @author Chris Lin
 * @version 1.0
 */
public class BinaryTreePostorderTraversal_145 {
    // Recursive version (LRD)
    // Time Complexity: O(n)
    // Space Complexity: O(n)
    // Runtime: 0 ms, faster than 100.00%
    // Memory Usage: 34.7 MB, less than 100.00%
    private List<Integer> mAnswer = new ArrayList<>();

    public List<Integer> postorderTraversal(TreeNode root) {
        if (root == null) {
            return mAnswer;
        }
        postorderTraversal(root.left);
        postorderTraversal(root.right);
        mAnswer.add(root.val);

        return mAnswer;
    }

    // Iterative version (LRD)
    // Time Complexity: O(n)
    // Space Complexity: O(n)
    // Runtime: 1 ms, faster than 61.18%
    // Memory Usage: 34.9 MB, less than 100.00%
    public List<Integer> postorderTraversal2(TreeNode root) {
        Stack<TreeNode> mStack = new Stack<>();
        mStack.push(root);
        while (!mStack.isEmpty()) {
            TreeNode mNode = mStack.pop();
            if (mNode == null) {
                continue;
            }
            mAnswer.add(mNode.val);
            mStack.push(mNode.left);
            mStack.push(mNode.right);
        }
        Collections.reverse(mAnswer);

        return mAnswer;
    }
}

File: Golang_src/medium/binaryTreePostorderTraversal_145.go
package medium

// Use recursive:
// Time Complexity: O(n)
// Space Complexity:O(n)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 2.1 MB, less than 36.18%
func postorderTraversal(root *TreeNode) []int {
	if root == nil {
		return nil
	}
	postOrder := []int{}
	postOrder = append(postOrder, postorderTraversal(root.Left)...)
	postOrder = append(postOrder, postorderTraversal(root.Right)...)
	postOrder = append(postOrder, root.Val)

	return postOrder
}

// Use iterative:
// Time Complexity: O(n)
// Space Complexity:O(n)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 2.1 MB, less than 36.18%
func postorderTraversal2(root *TreeNode) []int {
	result := []int{}
	stack := []*TreeNode{root}

	for len(stack) > 0 {
		node := stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		if node == nil {
			continue
		}
		result = append(result, node.Val)
		stack = append(stack, node.Left)
		stack = append(stack, node.Right)
	}
	size := len(result)
	for i := 0; i < size/2; i++ {
		result[i], result[size-1-i] = result[size-1-i], result[i]
	}

	return result
}
